url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/core/CloudSim.java

```diff
@@ -0,0 +1,928 @@
+		switch (e.getType()) {
+			case SimEvent.ENULL:
+				throw new IllegalArgumentException("Event has a null type.");
+
+			case SimEvent.CREATE:
+				SimEntity newe = (SimEntity) e.getData();
+				addEntityDynamically(newe);
+				break;
+
+			case SimEvent.SEND:
+				// Check for matching wait
+				dest = e.getDestination();
+				if (dest < 0) {
+					throw new IllegalArgumentException("Attempt to send to a null entity detected.");
+				} else {
+					int tag = e.getTag();
+					dest_ent = entities.get(dest);
+					if (dest_ent.getState() == SimEntity.WAITING) {
+						Integer destObj = Integer.valueOf(dest);
+						Predicate p = waitPredicates.get(destObj);
+						if ((p == null) || (tag == 9999) || (p.match(e))) {
+							dest_ent.setEventBuffer((SimEvent) e.clone());
+							dest_ent.setState(SimEntity.RUNNABLE);
+							waitPredicates.remove(destObj);
+						} else {
+							deferred.addEvent(e);
+						}
+					} else {
+						deferred.addEvent(e);
+					}
+				}
+				break;
+
+			case SimEvent.HOLD_DONE:
+				src = e.getSource();
+				if (src < 0) {
+					throw new IllegalArgumentException("Null entity holding.");
+				} else {
+					entities.get(src).setState(SimEntity.RUNNABLE);
+				}
+				break;
+
+			default:
+				break;
+		}
+	}
+
+	/**
+	 * Internal method used to start the simulation. This method should <b>not</b> be used by user
+	 * simulations.
+	 */
+	public static void runStart() {
+		running = true;
+		// Start all the entities
+		for (SimEntity ent : entities) {
+			ent.startEntity();
+		}
+
+		printMessage("Entities started.");
+	}
+
+	/**
+	 * Check if the simulation is still running. This method should be used by entities to check if
+	 * they should continue executing.
+	 * 
+	 * @return if the simulation is still running, otherwise
+	 */
+	public static boolean running() {
+		return running;
+	}
+
+	/**
+	 * This method is called if one wants to pause the simulation.
+	 * 
+	 * @return true, if successful otherwise.
+	 */
+	public static boolean pauseSimulation() {
+		paused = true;
+		return paused;
+	}
+
+	/**
+	 * This method is called if one wants to pause the simulation at a given time.
+	 * 
+	 * @param time the time at which the simulation has to be paused
+	 * @return true, if successful otherwise.
+	 */
+	public static boolean pauseSimulation(long time) {
+		if (time <= clock) {
+			return false;
+		} else {
+			pauseAt = time;
+		}
+		return true;
+	}
+
+	/**
+	 * This method is called if one wants to resume the simulation that has previously been paused.
+	 * 
+	 * @return if the simulation has been restarted or or otherwise.
+	 */
+	public static boolean resumeSimulation() {
+		paused = false;
+
+		if (pauseAt <= clock) {
+			pauseAt = -1;
+		}
+
+		return !paused;
+	}
+
+	/**
+	 * Start the simulation running. This should be called after all the entities have been setup
+	 * and added, and their ports linked.
+	 * 
+	 * @return the double last clock value
+	 */
+	public static double run() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.